lib/sdam/topology_description.js

```diff
@@ -0,0 +1,352 @@
+'use strict';
+const assert = require('assert');
+const ServerType = require('./server_description').ServerType;
+const ServerDescription = require('./server_description').ServerDescription;
+const ReadPreference = require('../topologies/read_preference');
+
+// contstants related to compatability checks
+const MIN_SUPPORTED_SERVER_VERSION = '2.6';
+const MIN_SUPPORTED_WIRE_VERSION = 2;
+const MAX_SUPPORTED_WIRE_VERSION = 5;
+
+// An enumeration of topology types we know about
+const TopologyType = {
+  Single: 'Single',
+  ReplicaSetNoPrimary: 'ReplicaSetNoPrimary',
+  ReplicaSetWithPrimary: 'ReplicaSetWithPrimary',
+  Sharded: 'Sharded',
+  Unknown: 'Unknown'
+};
+
+// Representation of a deployment of servers
+class TopologyDescription {
+  /**
+   * Create a TopologyDescription
+   *
+   * @param {string} topologyType
+   * @param {Map<string, ServerDescription>} serverDescriptions the a map of address to ServerDescription
+   * @param {string} setName
+   * @param {number} maxSetVersion
+   * @param {ObjectId} maxElectionId
+   */
+  constructor(topologyType, serverDescriptions, setName, maxSetVersion, maxElectionId, options) {
```

Comment:
We should POJO this constructor.

Reply:
I feel like this request keeps coming up in code reviews, is there a particular reason you want to do that beyond readability? This is an internal type to boot, so I'm not sure what it buys us.

Reply:
I dunno, I'm not convinced. Should we make a class called `TopologyUpdateParameters`? I think it's a little overkill just for the benefit of having named parameters. The named parameters example in that article is also not very compelling - agreed, definitely don't use "magic" boolean parameters. In this case, we are very clearly passing in named variables, essentially giving us the same benefit. On top of that, requiring an extra allocation of an object just to call a function seems needlessly wasteful.  I we can discuss POJOs for the future, but I don't think this PR is the place to make that choice